# NBA Player dataset
data: https://www.kaggle.com/justinas/nba-players-data

In [2]:
# modules
import pandas as pd
import numpy as np
from matplotlib import pyplot as plt

dataset_root = r"D:\\datasets\\csv"
dataset_name = r"\\nba\\all_seasons.csv"
dataset = dataset_root + dataset_name
df = pd.read_csv(dataset)
df

Unnamed: 0         player_name team_abbreviation   age  player_height  \
0               0       Dennis Rodman               CHI  36.0         198.12   
1               1   Dwayne Schintzius               LAC  28.0         215.90   
2               2        Earl Cureton               TOR  39.0         205.74   
3               3         Ed O'Bannon               DAL  24.0         203.20   
4               4         Ed Pinckney               MIA  34.0         205.74   
...           ...                 ...               ...   ...            ...   
11140       11140         Maxi Kleber               DAL  28.0         208.28   
11141       11141  Melvin Frazier Jr.               ORL  23.0         195.58   
11142       11142      Meyers Leonard               MIA  28.0         213.36   
11143       11143        Norvel Pelle               PHI  27.0         208.28   
11144       11144         Matt Thomas               TOR  25.0         193.04   

       player_weight                      college  country draft_year  \
0          99.790240  Southeastern Oklahoma State      USA       1986   
1         117.933920                      Florida      USA       1990   
2          95.254320                Detroit Mercy      USA       1979   
3         100.697424                         UCLA      USA       1995   
4         108.862080                    Villanova      USA       1985   
...              ...                          ...      ...        ...   
11140     108.862080                         None  Germany  Undrafted   
11141      97.522280                       Tulane      USA       2018   
11142     117.933920                     Illinois      USA       2012   
11143     104.779752                         None      USA  Undrafted   
11144      86.182480                   Iowa State      USA  Undrafted   

      draft_round  ...  pts   reb  ast  net_rating  oreb_pct  dreb_pct  \
0               2  ...  5.7  16.1  3.1        16.1     0.186     0.323   
1               1  ...  2.3   1.5  0.3        12.3     0.078     0.151   
2               3  ...  0.8   1.0  0.4        -2.1     0.105     0.102   
3               1  ...  3.7   2.3  0.6        -8.7     0.060     0.149   
4               1  ...  2.4   2.4  0.2       -11.2     0.109     0.179   
...           ...  ...  ...   ...  ...         ...       ...       ...   
11140   Undrafted  ...  9.1   5.4  1.1         4.6     0.056     0.140   
11141           2  ...  1.2   0.3  0.1        -2.4     0.018     0.058   
11142           1  ...  6.1   5.1  1.1         5.6     0.029     0.217   
11143   Undrafted  ...  2.1   3.0  0.4       -16.4     0.085     0.237   
11144   Undrafted  ...  4.5   1.4  0.5         1.0     0.017     0.104   

       usg_pct  ts_pct  ast_pct   season  
0        0.100   0.479    0.113  1996-97  
1        0.175   0.430    0.048  1996-97  
2        0.103   0.376    0.148  1996-97  
3        0.167   0.399    0.077  1996-97  
4        0.127   0.611    0.040  1996-97  
...        ...     ...      ...      ...  
11140    0.136   0.605    0.064  2019-20  
11141    0.164   0.480    0.033  2019-20  
11142    0.120   0.640    0.076  2019-20  
11143    0.126   0.521    0.056  2019-20  
11144    0.149   0.663    0.089  2019-20  

[11145 rows x 22 columns]

1-) Get first 10 records.

In [3]:
# -*-* code *-*-
df.head(10)

Unnamed: 0        player_name team_abbreviation   age  player_height  \
0           0      Dennis Rodman               CHI  36.0         198.12   
1           1  Dwayne Schintzius               LAC  28.0         215.90   
2           2       Earl Cureton               TOR  39.0         205.74   
3           3        Ed O'Bannon               DAL  24.0         203.20   
4           4        Ed Pinckney               MIA  34.0         205.74   
5           5      Eddie Johnson               HOU  38.0         200.66   
6           6        Eddie Jones               LAL  25.0         198.12   
7           7     Elden Campbell               LAL  28.0         213.36   
8           8  Eldridge Recasner               ATL  29.0         193.04   
9           9       Elliot Perry               MIL  28.0         182.88   

   player_weight                      college country draft_year draft_round  \
0      99.790240  Southeastern Oklahoma State     USA       1986           2   
1     117.933920                      Florida     USA       1990           1   
2      95.254320                Detroit Mercy     USA       1979           3   
3     100.697424                         UCLA     USA       1995           1   
4     108.862080                    Villanova     USA       1985           1   
5      97.522280                     Illinois     USA       1981           2   
6      86.182480                       Temple     USA       1994           1   
7     113.398000                      Clemson     USA       1990           1   
8      86.182480                   Washington     USA       1992   Undrafted   
9      72.574720                      Memphis     USA       1991           2   

   ...   pts   reb  ast  net_rating  oreb_pct  dreb_pct  usg_pct  ts_pct  \
0  ...   5.7  16.1  3.1        16.1     0.186     0.323    0.100   0.479   
1  ...   2.3   1.5  0.3        12.3     0.078     0.151    0.175   0.430   
2  ...   0.8   1.0  0.4        -2.1     0.105     0.102    0.103   0.376   
3  ...   3.7   2.3  0.6        -8.7     0.060     0.149    0.167   0.399   
4  ...   2.4   2.4  0.2       -11.2     0.109     0.179    0.127   0.611   
5  ...   8.2   2.7  1.0         4.1     0.034     0.126    0.220   0.541   
6  ...  17.2   4.1  3.4         4.1     0.035     0.091    0.209   0.559   
7  ...  14.9   8.0  1.6         3.3     0.095     0.183    0.222   0.520   
8  ...   5.7   1.6  1.3        -0.3     0.036     0.076    0.172   0.539   
9  ...   6.9   1.5  3.0        -1.2     0.018     0.081    0.177   0.557   

   ast_pct   season  
0    0.113  1996-97  
1    0.048  1996-97  
2    0.148  1996-97  
3    0.077  1996-97  
4    0.040  1996-97  
5    0.102  1996-97  
6    0.149  1996-97  
7    0.087  1996-97  
8    0.141  1996-97  
9    0.262  1996-97  

[10 rows x 22 columns]

2-) How many records are there?

In [9]:
# -*-* code *-*-
len(df.index)

11145

3-) What is the average height of all players?

In [12]:
# -*-* code *-*-
df["player_height"].mean()

200.81281830417785

4-) What is the maximum height of all players?

In [13]:
# -*-* code *-*-
df["player_height"].max()
# df["player_height"].min()  # min

231.14

5-) Who is the player with the highest height?

In [24]:
# -*-* code *-*-
mask = df["player_height"] == df["player_height"].max()
longest_player = df[mask]["player_name"]
print(longest_player)
print(longest_player.iloc[0])

92      Gheorghe Muresan
1212    Gheorghe Muresan
1408    Gheorghe Muresan
Name: player_name, dtype: object
Gheorghe Muresan


6-) List player names and teams between the ages of 20-25 and sort in descending order.

In [37]:
# -*-* code *-*-
mask_age = (df["age"]>=20) & (df["age"] < 25)
indexes = ["player_name", "team_abbreviation", "age"]
filtered = df[mask_age][indexes]
filtered.sort_values("age", ascending=False)  # descending order.

player_name team_abbreviation   age
3                   Ed O'Bannon               DAL  24.0
8058                 Jan Vesely               DEN  24.0
3797                 Brian Cook               LAL  24.0
3793           Curtis Borchardt               UTA  24.0
3791                  DJ Mbenga               DAL  24.0
...                         ...               ...   ...
10731                 Ja Morant               MEM  20.0
4440             Ersan Ilyasova               MIL  20.0
3986   Martynas Andriuskevicius               CLE  20.0
3985            Marvin Williams               ATL  20.0
5794           Anthony Randolph               GSW  20.0

[3531 rows x 3 columns]

7-) What team does "John Holland" play?

In [41]:
# -*-* code *-*-
name = "John Holland"
mask_name = df["player_name"] == name
df[mask_name]["team_abbreviation"].iloc[0]

'CLE'

8-) What is the average height of the players by teams?

In [42]:
# -*-* code *-*-
# "by" -> pandas groupby
df.groupby("team_abbreviation").mean()["player_height"]

team_abbreviation
ATL    201.170653
BKN    200.569286
BOS    199.960981
CHA    200.325669
CHH    200.916854
CHI    200.391545
CLE    200.920513
DAL    201.261927
DEN    200.447796
DET    200.184637
GSW    200.710000
HOU    200.816349
IND    201.871806
LAC    200.104987
LAL    201.550027
MEM    200.145270
MIA    200.961912
MIL    201.078544
MIN    200.444863
NJN    201.313619
NOH    200.961958
NOK    200.025000
NOP    200.246393
NYK    201.474717
OKC    200.818051
ORL    200.418424
PHI    200.902842
PHX    201.059945
POR    201.339401
SAC    200.787709
SAS    200.297082
SEA    201.650879
TOR    200.972923
UTA    200.737611
VAN    199.531111
WAS    201.062111
Name: player_height, dtype: float64

9-) How many different teams are there?

In [46]:
# -*-* code *-*-
# print(len(df.groupby("team_abbreviation")))
print(df["team_abbreviation"].nunique())  # second way.

36


10-) How many players are there in each team?

In [47]:
# -*-* code *-*-
df["team_abbreviation"].value_counts()

TOR    390
CLE    390
LAC    389
MIA    387
DAL    384
ATL    383
PHI    380
WAS    379
HOU    378
SAS    377
GSW    376
DEN    372
MIL    371
IND    371
NYK    371
LAL    371
CHI    369
ORL    368
POR    367
BOS    367
MIN    366
PHX    362
UTA    360
SAC    358
DET    358
MEM    315
NJN    257
CHA    254
OKC    195
SEA    182
NOH    143
BKN    140
NOP    122
CHH     89
VAN     72
NOK     32
Name: team_abbreviation, dtype: int64

11.1-) Find the records containing the word "and"
Note: only lowercases

In [62]:
# -*-* code *-*-
df.dropna(inplace=True)  # drop "na" values
mask_contain = df["player_name"].str.contains("and")
df[mask_contain]

Unnamed: 0               player_name team_abbreviation   age  \
22             22          Erick Strickland               DAL  23.0   
145           145     Aleksandar Djordjevic               POR  29.0   
171           171            Cory Alexander               SAS  24.0   
237           237          Shandon Anderson               UTA  23.0   
250           250        Randolph Childress               DET  24.0   
...           ...                       ...               ...   ...   
10955       10955            Tyson Chandler               HOU  37.0   
10972       10972           Wilson Chandler               BKN  32.0   
11038       11038             Landry Shamet               LAC  22.0   
11057       11057             Julius Randle               NYK  25.0   
11102       11102  Nickeil Alexander-Walker               NOP  21.0   

       player_height  player_weight        college country draft_year  \
22            190.50      95.254320       Nebraska     USA  Undrafted   
145           187.96      89.811216           None     USA  Undrafted   
171           185.42      86.182480       Virginia     USA       1995   
237           198.12      94.347136        Georgia     USA       1996   
250           187.96      85.275296    Wake Forest     USA       1995   
...              ...            ...            ...     ...        ...   
10955         213.36     106.594120           None     USA       2001   
10972         203.20     106.594120         DePaul     USA       2007   
11038         193.04      86.182480  Wichita State     USA       2018   
11057         203.20     113.398000       Kentucky     USA       2014   
11102         195.58      92.986360  Virginia Tech  Canada       2019   

      draft_round  ...   pts  reb  ast  net_rating  oreb_pct  dreb_pct  \
22      Undrafted  ...  10.6  3.2  2.4        -6.4     0.032     0.112   
145     Undrafted  ...   3.1  0.6  0.6         4.8     0.021     0.070   
171             1  ...   7.2  1.5  3.2        -8.8     0.023     0.081   
237             2  ...   5.9  2.8  0.8         2.5     0.065     0.139   
250             1  ...   1.7  0.3  0.7        -0.9     0.008     0.037   
...           ...  ...   ...  ...  ...         ...       ...       ...   
10955           1  ...   1.3  2.5  0.2         5.7     0.097     0.172   
10972           1  ...   5.9  4.1  1.1        -8.3     0.015     0.164   
11038           1  ...   9.9  2.0  1.8         5.0     0.004     0.059   
11057           1  ...  19.3  9.7  3.1        -5.7     0.067     0.223   
11102           1  ...   5.1  2.0  1.8        -5.1     0.015     0.135   

       usg_pct  ts_pct  ast_pct   season  
22       0.216   0.510    0.161  1996-97  
145      0.179   0.687    0.135  1996-97  
171      0.218   0.528    0.300  1996-97  
237      0.193   0.534    0.074  1996-97  
250      0.176   0.448    0.205  1996-97  
...        ...     ...      ...      ...  
10955    0.057   0.717    0.037  2019-20  
10972    0.130   0.529    0.078  2019-20  
11038    0.129   0.611    0.085  2019-20  
11057    0.270   0.532    0.160  2019-20  
11102    0.227   0.441    0.207  2019-20  

[361 rows x 22 columns]

11.2-) Find the records containing the word "and"
Note: ignore uppercase and lowercase

In [67]:
# -*-* code *-*-
df.dropna(inplace=True)  # drop "na" values

def str_lower_find(name, ext):
    return ext in name.lower()

mask_contain = df["player_name"].apply(func=str_lower_find, args=("and",))
df[mask_contain]

Unnamed: 0               player_name team_abbreviation   age  \
22             22          Erick Strickland               DAL  23.0   
97             97             Greg Anderson               SAS  33.0   
145           145     Aleksandar Djordjevic               POR  29.0   
151           151           Andrew DeClercq               GSW  24.0   
152           152               Andrew Lang               MIL  31.0   
...           ...                       ...               ...   ...   
11038       11038             Landry Shamet               LAC  22.0   
11055       11055     Juan Toscano-Anderson               GSW  26.0   
11057       11057             Julius Randle               NYK  25.0   
11058       11058           Justin Anderson               BKN  26.0   
11102       11102  Nickeil Alexander-Walker               NOP  21.0   

       player_height  player_weight        college country draft_year  \
22            190.50      95.254320       Nebraska     USA  Undrafted   
97            208.28     113.398000        Houston     USA       1987   
145           187.96      89.811216           None     USA  Undrafted   
151           208.28     104.326160        Florida     USA       1995   
152           210.82     124.737800       Arkansas     USA       1988   
...              ...            ...            ...     ...        ...   
11038         193.04      86.182480  Wichita State     USA       2018   
11055         198.12      94.800728           None     USA  Undrafted   
11057         203.20     113.398000       Kentucky     USA       2014   
11058         195.58     104.779752       Virginia     USA       2015   
11102         195.58      92.986360  Virginia Tech  Canada       2019   

      draft_round  ...   pts  reb  ast  net_rating  oreb_pct  dreb_pct  \
22      Undrafted  ...  10.6  3.2  2.4        -6.4     0.032     0.112   
97              1  ...   3.9  5.5  0.4       -14.2     0.109     0.219   
145     Undrafted  ...   3.1  0.6  0.6         4.8     0.021     0.070   
151             2  ...   5.3  4.2  0.5        -5.6     0.132     0.198   
152             2  ...   5.3  5.3  0.5        -1.2     0.096     0.182   
...           ...  ...   ...  ...  ...         ...       ...       ...   
11038           1  ...   9.9  2.0  1.8         5.0     0.004     0.059   
11055   Undrafted  ...   5.8  4.3  2.2       -12.6     0.031     0.163   
11057           1  ...  19.3  9.7  3.1        -5.7     0.067     0.223   
11058           1  ...   1.0  0.7  0.0       -25.4     0.000     0.091   
11102           1  ...   5.1  2.0  1.8        -5.1     0.015     0.135   

       usg_pct  ts_pct  ast_pct   season  
22       0.216   0.510    0.161  1996-97  
97       0.106   0.531    0.033  1996-97  
145      0.179   0.687    0.135  1996-97  
151      0.173   0.552    0.050  1996-97  
152      0.123   0.498    0.031  1996-97  
...        ...     ...      ...      ...  
11038    0.129   0.611    0.085  2019-20  
11055    0.124   0.537    0.145  2019-20  
11057    0.270   0.532    0.160  2019-20  
11058    0.159   0.218    0.000  2019-20  
11102    0.227   0.441    0.207  2019-20  

[650 rows x 22 columns]